### PR TITLE
Add support for Security CommandClass and Door Lock

### PR DESCRIPTION
**Pull request by snagytx**
Add SecurityHandler for Security CommandClass and any required changes to make DoorLocks work. Tested with Schlage BE469

There are required a few changes at the main components, I tried to keep them at the minimum, which also caused required me to add SecurityHandler outside of Security CommandClass and make all the required calls from Security CommandClass to SecurityHandler.

Tested compilation of Windows and Ubuntu 14 - On Ubuntu I also creates a deb file and deployed on a R Pi - on windows I didn't create an installer - so I don't know if that works or not - but from the debugger everything is deployed correctly.

Maybe you guys can implement it in a better way because you know the internals better - if needed you can send me the code and I'll test it.

Let me know if you have any questions.

### CODE REVIEW

File: BaseFiles/Common/Common.csproj [hunk 1]
```diff
@@ -4538,13 +4544,5 @@
...
   <ItemGroup />
-  <ItemGroup>
-    <Folder Include="html\pages\control\widgets\homegenie\generic\locales\" />
-    <Folder Include="html\js\i18n\" />
-    <Folder Include="html\pages\configure\widgeteditor\" />
-    <Folder Include="html\js\codemirror-4.03\mode\css\" />
-    <Folder Include="html\js\codemirror-4.03\mode\htmlembedded\" />
-    <Folder Include="html\js\codemirror-4.03\mode\htmlmixed\" />
-    <Folder Include="html\js\codemirror-4.03\mode\xml\" />
-  </ItemGroup>
```
Comment (genielabs) [inline BaseFiles/Common/Common.csproj]: I am not sure why did you remove these entries?

File: BaseFiles/Common/html/pages/control/widgets/homegenie/generic/doorlock.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+﻿[{
+    Name: "Security Alarm System",
+    Author: "Generoso Martello",
```
Comment (genielabs) [inline BaseFiles/Common/html/pages/control/widgets/homegenie/generic/doorlock.js]: @snagytx  Here should appear your name =)

Comment (snagytx) [inline BaseFiles/Common/html/pages/control/widgets/homegenie/generic/doorlock.js]: true, there are many places where I should have done something different. I had the code in a rough version since revision 372 - close to a half a year - but didn't get a chance to finish it and also it took me a few days to adapt the code to the new changes - I didn't want to wait more and then have to catch-up again :)

File: BaseFiles/Common/html/pages/events/_events.js [hunk 3]
```diff
@@ -263,13 +263,20 @@ HG.WebApp.Events.SendEventToUi = function (module, eventLog) {
...
                     var propname = eventLog.Property.substring(eventLog.Property.indexOf('.') + 1);
-                    var value = (parseFloat(eventLog.Value.replace(',', '.')).toFixed(2));
-                    // TODO: should level be reported as is??
-                    if (propname == 'Level') {
-                        value = value * 100;
-                        if (value > 98) value = 100;
-                    }
-                    value += '%';
+                    
+                    if(propname == "DoorLock"){
```
Comment (genielabs) [inline BaseFiles/Common/html/pages/events/_events.js]: (self-note) instead of adding any other specific code, will add checking of eventLog.Value type, and if it is numeric it will be processed (eg. toFixed), otherwise it will be left as is.

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs [hunk 4]
```diff
@@ -337,9 +341,17 @@ private void ZwaveMessageReceived(object sender, ZWaveMessageReceivedEventArgs a
...
+                                byte[] nodeInfo = new byte[nodeInfoLength];
+                                Array.Copy(args.Message, 8, nodeInfo, 0, nodeInfoLength);
                                 newNode.NodeInformationFrame = nodeInfo;
+
+                                newNode.BasicClass = args.Message[8];
+                                newNode.GenericClass = args.Message[9];
+                                newNode.SpecificClass = args.Message[10];
+                                devices.Add(newNode);
+                                newNode.HandleSecureCommandClasses(nodeInfo);
+
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: (self-note)chek this out

File: BaseFiles/Common/Common.csproj (hunk 1 above)
Comment (snagytx) [inline BaseFiles/Common/Common.csproj]: they were removed by mistake - re-added them in a later commit

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs (hunk 4 above)
Comment (snagytx) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: when a code is included we need to do a handshake and set the new security key - That has to happen at the inclusion time.

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs [hunk 5]
```diff
@@ -443,13 +455,21 @@ private void ZwaveMessageReceived(object sender, ZWaveMessageReceivedEventArgs a
...
+                                byte[] nodeInfo = new byte[nifLength];
                                 //Console.WriteLine(ByteArrayToString(args.Message));
-                                Array.Copy(args.Message, 7, nodeInfo, 0, nifLength - 2);
+                                Array.Copy(args.Message, 7, nodeInfo, 0, nifLength);
                                 znode.NodeInformationFrame = nodeInfo;
+                                var triggerEvents = znode.HandleNodeUpdate(args.Message);
                                 //
-                                RaiseUpdateParameterEvent(new ZWaveEvent(znode, EventParameter.NodeInfo, Utility.ByteArrayToString(nodeInfo), 0));
-                                RaiseUpdateParameterEvent(new ZWaveEvent(znode, EventParameter.WakeUpNotify, "1", 0));
+                                if (!triggerEvents) {
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: @snagytx should this be withouth the '!' modifier, or perhaps this is meant to have a different name?

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs [hunk 6]
```diff
@@ -661,6 +681,15 @@ private void LoadNodesConfig()
                 var serializer = new XmlSerializer(nodesConfig.GetType());
                 var reader = new StreamReader(configPath);
                 nodesConfig = (List<ZWaveNodeConfig>)serializer.Deserialize(reader);
+                foreach (ZWaveNodeConfig node in nodesConfig) {
+                    var newNode = CreateDevice(node.NodeId, 0x00);
+                    newNode.NodeInformationFrame = node.NodeInformationFrame;
+                    newNode.SecuredNodeInformationFrame = node.SecuredNodeInformationFrame;
+                    devices.Add(newNode);
+
+                    newNode.BuildSupportedList(node.NodeInformationFrame, false);
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: (self-note)check this out

File: MigFiles/SupportLibraries/ZWaveLib/Enums/SecurityCommand.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace ZWaveLib
+{
+    public enum SecurityCommand
+    {
+        COMMAND_SUPPORTED_GET = 0x02,
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/Enums/SecurityCommand.cs]: @snagytx please adopt same naming convention used in similar files

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs (hunk 5 above)
Comment (snagytx) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: I think the name is incorrect HandleNodeUpdate returns true if a this security CommandClass is supported

File: MigFiles/SupportLibraries/ZWaveLib/ZWaveLib.csproj [hunk 8]
```diff
@@ -95,7 +100,5 @@
     </ProjectReference>
   </ItemGroup>
   <ItemGroup />
-  <ItemGroup>
-    <Folder Include="Enums\" />
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/ZWaveLib.csproj]: @snagytx  I'm not sure why this folder remove here too...

File: MigFiles/SupportLibraries/ZWaveLib/ZWaveNode.cs [hunk 9]
```diff
@@ -27,29 +27,85 @@
...
+using System.Diagnostics;
+
 using ZWaveLib.CommandClasses;
 using ZWaveLib.Values;
 
 namespace ZWaveLib
 {
 
+    public class SupportedCommandClasses : IEquatable<SupportedCommandClasses>
+    {
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/ZWaveNode.cs]: (self-note)check this and other mods. perhaps some code can be moved out of here.

File: MigFiles/SupportLibraries/ZWaveLib/ZWavePort.cs [hunk 10]
```diff
@@ -123,6 +123,7 @@ public byte SendMessage(ZWaveMessage message, bool disableCallback = false)
                 if (disableCallback)
                 {
                     message.Message[message.Message.Length - 1] = GenerateChecksum(message.Message);
+                    Utility.logMessage("    Sent1   : " + Utility.ByteArrayToString(message.Message));
```
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/ZWavePort.cs]: @snagytx please remove these Utility.logMessage (the function as well), because there's already debug output from serial port.

**Comment by genielabs** [comment]
do all the fix you can. Let me know when finished so I will merge and work on the new code (see self-notes).

File: MigFiles/SupportLibraries/ZWaveLib/ZWavePort.cs (hunk 10 above)
Comment (snagytx) [inline MigFiles/SupportLibraries/ZWaveLib/ZWavePort.cs]: Using the same function for Security debugging

File: MigFiles/SupportLibraries/ZWaveLib/Controller.cs (hunk 4 above)
Comment (genielabs) [inline MigFiles/SupportLibraries/ZWaveLib/Controller.cs]: ok.

**Comment by snagytx** [comment]
I revisited through all the changes and cleaned up the code - keeping in mind your comments. See if there is any requirement for more changes

**Comment by genielabs** [comment]
thanks, I'll take care of other minor changes.

**Comment by genielabs** [comment]
hi @snagytx , this is going to be much more complex than I imagined, so I need your help.
I'm now going to commit some work-in-progress around your code.
Mainly I've been trying to make it fit into the current library structure, so I had to revert some changes you made mainly to ZWaveNode and tried to put the pieces of code in a better place.
These modifications, broke up your code and actually it won't even compile.
I'm pushing this the same, just to give an idea of how code should be arranged.
The main work has to be done around the Security.cs file and Controller.cs file I guess.
Just few notes about file implementing ICommandClass interface:
- these should have most methods as static (but the GetEvent) called by passing the reference to the node
- when it needs to store some data into a node, use the ZWaveNode.Data dictionary (see GetSchemeAgreedData and SetSchemeAgreedData example in Security.cs file)
Still if you figure out some better solution that could be applied as a generic solution, please feel free to suggest it.
Let me know your thoughts about this.

**Comment by genielabs** [comment]
@snagytx this is the commit: https://github.com/genielabs/HomeGenie/commit/a955d81c5f156cfb352c16b439e8f547546907c5
